hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java

```diff
@@ -2241,11 +2248,18 @@ public SessionFactoryImplementor getSessionFactory() {
 	public void initializeCollection(PersistentCollection collection, boolean writing) {
 		checkOpenOrWaitingForAutoClose();
 		checkTransactionSynchStatus();
-		InitializeCollectionEvent event = new InitializeCollectionEvent( collection, this );
-		for ( InitializeCollectionEventListener listener : listeners( EventType.INIT_COLLECTION ) ) {
-			listener.onInitializeCollection( event );
+		boolean success = false;
+		try {
+			InitializeCollectionEvent event = new InitializeCollectionEvent( collection, this );
+			for ( InitializeCollectionEventListener listener : listeners( EventType.INIT_COLLECTION ) ) {
+				listener.onInitializeCollection( event );
+			}
+			success = true;
+		}
+		finally {
+			delayedAfterCompletion();
+			afterOperation( success );
```

Comment:
When initializing an entity and its collections in one go, we will now call `afterOperation()` several times (after each collection initialization). It won't have any consequence (e.g. it won't close the connection or close the transaction and reopen a new one for the next collection)?

Reply:
I think what could happen is that after the connection is closed any other operation will cause an exception due to the closed connection. 
I'm going to create a test for the scenario you proposed.